test/puppetlabs/puppetdb/generative/generators.clj

```diff
@@ -0,0 +1,283 @@
+
+(defgen :resource/type (string+))
+(defgen :resource/title (string+))
+(defgen :resource/exported gen/boolean)
+(defgen :resource/file (string+ 1024))
+(defgen :resource/line pg-pos-integer)
+
+(defgen :resource/tag  (gen/string-from-regex #"[a-z0-9_][a-z0-9_:\-.]*"))
+(defgen :resource/tags (gen/vector :resource/tag 0 3))
+
+(defgen :resource.param/name (string+ 40))
+(defgen :resource.param/value
+  (gen/recursive-gen
+   (fn [inner-gen]
+     (gen/one-of [(gen/map (string+) inner-gen)
+                  (gen/vector inner-gen)]))
+   (gen/one-of [(string+)
+                pg-smallint
+                gen/boolean])))
+
+(defgen :resource/parameters (gen/map :resource.param/name :resource.param/value))
+(defgen :catalog/resource (gen/keys :resource/type
+                                    :resource/title
+                                    :resource/exported
+                                    :resource/file
+                                    :resource/line
+                                    :resource/parameters
+                                    :resource/tags))
+
+(defgen :catalog/resources
+  (gen/vector-distinct :catalog/resource))
+
+(defgen :catalog/resource-tree
+  (gen/recursive-gen
+   (fn [inner-gen]
+     (gen/hash-map :children (gen/vector inner-gen)
+                   :resource :catalog/resource
+                   :relationship :edge/relationship))
+   (gen/keys :catalog/resource)))
+
+(defn resource-tree-to-edge-list [rt]
+  (let [nodes (tree-seq :children :children rt)
+        resources (map :resource nodes)
+        edges (->> nodes
+                   (filter :children)
+                   (mapcat (fn [node]
+                             (let [source (:resource node)
+                                   target-nodes (:children node)
+                                   relationship (:relationship node)]
+                               (map (fn [target-node]
+                                      {:source (select-keys source [:type :title])
+                                       :target (select-keys (:resource target-node) [:type :title])
+                                       :relationship relationship})
+                                    target-nodes)))))]
+    {:resources resources
+     :edges edges}))
+
+(defgen :puppet/catalog
+  (gen/fmap (fn [cat]
+              (-> cat
+                  (dissoc :resource-tree)
+                  (merge (resource-tree-to-edge-list (:resource-tree cat)))))
+            (gen/keys :puppet/certname
+                      :puppet/environment
+                      :puppet/transaction_uuid
+                      :puppet/catalog_uuid
+                      :puppet/producer_timestamp
+                      :puppet/producer
+                      :catalog/version
+                      :catalog/code_id
+                      :catalog/resource-tree)))
+
+;;; factset generator
+
+(defgen :fact/name (string+))
+(defgen :fact/value json-value)
```

Comment:
Will this only generate vectors and maps? Or will it generate strings/integers as well (not inside a tree structure)